torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -562,39 +573,44 @@ def moe_infer(self, x, topk_ids, topk_weight):
             dist.all_to_all_single(
                 tokens_per_expert_group, tokens_per_expert, group=self.ep_group
             )
-            torch.sum(
-                tokens_per_expert_group.view(self.ep_size, -1),
-                dim=1,
-                out=self.output_splits,
-            )
-            # print(
-            #     f"EP rank {self.ep_rank} receives {self.output_splits} tokens"
-            #     f"EP rank {self.ep_rank} sends {self.input_splits} tokens"
-            # )
-            # Total received tokens of current rank
-            received = torch.empty(1, dtype=torch.int64, device=x.device)
-            # DP to EP token shuffle
-            on_device_all_to_all_v(
-                self.token_gather_buf,
-                received,
-                sorted_tokens,
-                self.input_splits,
-                group=self.ep_group,
-            )
+
+        # DP to EP token shuffle. This part needs gradient.
+        on_device_all_to_all_v(
+            self.token_gather_buf,
+            self.output_splits,
+            sorted_tokens,
+            self.input_splits,
+            self.ep_group,
+        )
+
+        with torch.no_grad():
+            # Output splits sanity check
+            # expected_splits = tokens_per_expert_group.view(self.ep_size, -1).sum(dim=1)
+            # torch.testing.assert_close(self.output_splits, expected_splits)
             # Received tokens from all other ranks. TODO: use mask instead
-            gathered_tokens = self.token_gather_buf[:received]
+            received = self.output_splits.sum()
+
+        # TODO: don't use `received`
+        gathered_tokens = self.token_gather_buf[:received]
+
+        # This part prepares the token indices for each expert, because we
+        # receive them in a whole chunk for the entire expert group (i.e.,
+        # device). This part doesn't need gradient.
+        with torch.no_grad():
             tokens_per_expert_post_gather = tokens_per_expert_group.view(
                 self.ep_size, self.experts_per_rank
             ).sum(dim=0)
             gatherd_idxs = np.zeros(shape=(gathered_tokens.shape[0],), dtype=np.int32)
             s = 0
-            for i, k in enumerate(tokens_per_expert_group.cpu().numpy()):
+            # TODO: remove `tolist()`
+            for i, k in enumerate(tokens_per_expert_group.tolist()):
```

Comment:
this is still a cuda-cpu sync point, right? can we avoid it by doing a scan on device or something?